src/java/org/apache/cassandra/io/storage/StorageProvider.java

```diff
@@ -0,0 +1,133 @@
+{
+    Logger logger = LoggerFactory.getLogger(StorageProvider.class);
+
+    String CUSTOM_STORAGE_PROVIDER = "cassandra.custom_storage_provider";
+    StorageProvider instance = System.getProperty(CUSTOM_STORAGE_PROVIDER) == null
+        ? new DefaultProvider()
+        : FBUtilities.instantiate(System.getProperty(CUSTOM_STORAGE_PROVIDER), "storage provider");
+
+    enum DirectoryType
+    {
+        DATA("data_file_directories"),
+        LOCAL_SYSTEM_DATA("local_system_data_file_directories"),
+        COMMITLOG("commit_log_directory"),
+        HINTS("hints_directory"),
+        SAVED_CACHES("saved_caches_directory"),
+        CDC("cdc_raw_directory"),
+        SNAPSHOT("snapshot_directory"),
+        NODES("nodes_local_directory"),
+        LOG_TRANSACTION("log_transaction_directory"),
+        LOGS("logs_directory"),
+        TEMP("temp_directory"),
+        OTHERS("other_directory");
+
+        final String what;
+
+        final boolean readable;
+        final boolean writable;
+
+        DirectoryType(String what)
+        {
+            this.what = what;
+            this.readable = true;
+            this.writable = true;
+        }
+    }
+
+    /**
+     * @return local path if given path is remote path, otherwise returns itself
+     */
+    File getLocalPath(File path);
+
+    /**
+     * update the given path with open options for the sstable components
+     */
+    File withOpenOptions(File ret, Component component);
+
+    /**
+     * Create data directories for given table
+     *
+     * @param ksMetadata The keyspace metadata, can be null. This is used when schema metadata is
+     *                   not available in {@link SchemaManager}, eg. CNDB backup & restore
+     * @param keyspaceName the name of the keyspace
+     * @param dirs current local data directories
+     * @return data directories that are created
+     */
+    Directories.DataDirectory[] createDataDirectories(@Nullable KeyspaceMetadata ksMetadata, String keyspaceName, Directories.DataDirectory[] dirs);
+
+    /**
+     * Create directory for the given path and type, either locally or remotely if any remote storage parameters are passed in.
+     *
+     * @param dir the directory path to create
+     * @param type the type of directory to create
+     * @return the actual directory path, which can be either local or remote; or null if directory can't be created
+     */
+    File createDirectory(String dir, DirectoryType type);
+
+    class DefaultProvider implements StorageProvider
+    {
+        @Override
+        public File getLocalPath(File path)
+        {
+            return path;
+        }
+
+        @Override
+        public File withOpenOptions(File ret, Component component)
+        {
+            return ret;
+        }
+
+        @Override
+        public Directories.DataDirectory[] createDataDirectories(@Nullable KeyspaceMetadata ksMetadata, String keyspaceName, Directories.DataDirectory[] dirs)
+        {
+            return dirs;
```

Comment:
why it does not create any directory

Reply:
For default implementation, they are `DatabaseDescriptor#dataDirectories`, already created in `createAllDirectories()`

Reply:
Maybe add a comment with that explanation?